File: packages/engine/src/cpus/6502/instructions/lax.js
import { CycleType } from '..'

/// extra opcode lax = lda + ldx
export function* lax(mode, addr) {
  this.acc = this.read(Array.isArray(addr) ? addr[0] : addr)
  this.x = this.acc
  this.zero = this.x === 0
  this.negative = (this.x & 0x80) > 0
  yield CycleType.GET
}

File: packages/engine/src/cores/famibob/cpu/instructions/lax.zig
const std = @import("std");
const CPU = @import("../cpu.zig").CPU;
const Addressing = @import("../opcode.zig").Addressing;

// extra opcode lax = lda + ldx
pub fn lax(self: *CPU, addressing: Addressing) void {
    var b: u8 = 0;

    switch (addressing) {
        .imm => |v| {
            b = v;
        },
        .zpg, .zpx, .zpy => |v| {
            b = self.read(v);
        },
        .abs => |v| {
            b = self.read(v);
        },
        .abx, .aby => |v| {
            b = self.read(v.@"0");
        },
        .idx => |v| {
            b = self.read(v.@"0");
        },
        .idy => |v| {
            b = self.read(v.@"0");
        },
        else => {},
    }

    self.acc = b;
    self.x = self.acc;
    self.z = self.x == 0;
    self.n = (self.x & 0x80) > 0;
}
